From b80c19afccb622c3ca3f2deae4997369035d6ab6 Mon Sep 17 00:00:00 2001
Message-Id: <b80c19afccb622c3ca3f2deae4997369035d6ab6.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Gaurav K Singh <gaurav.k.singh@intel.com>
Date: Wed, 14 Jan 2015 00:06:15 +0530
Subject: [PATCH 47/95] MUST_REBASE [VPG]: drivers/video/adf: Fix the issue of
 long packet send to the DSI Controller

Currently we are checking for fifo full. When the fifo full bit is cleared by
hardware there will be only space for 4 more bytes to be pushed and hence it
may not be enough to push long packet. In order to push long packets, make
sure that the generic data fifo is empty before sending long packet.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-5276
Change-Id: I1daf599f0799f08564db8deffefb44e3563047af
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
---
 .../adf/intel/core/common/dsi/intel_dsi_cmd.c      |   18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/drivers/video/adf/intel/core/common/dsi/intel_dsi_cmd.c b/drivers/video/adf/intel/core/common/dsi/intel_dsi_cmd.c
index 29554057..d6c5afa 100644
--- a/drivers/video/adf/intel/core/common/dsi/intel_dsi_cmd.c
+++ b/drivers/video/adf/intel/core/common/dsi/intel_dsi_cmd.c
@@ -164,7 +164,7 @@ static int adf_dsi_vc_send_long(struct dsi_pipe *dsi_pipe, int channel,
 	struct dsi_config *config = &dsi_pipe->config;
 	struct dsi_context *intel_dsi = &dsi_pipe->config.ctx;
 	int pipe = config->pipe;
-	u32 data_reg;
+	u32 data_reg, ctrl_reg, ctrl;
 	int i, j, n;
 	u32 mask;
 
@@ -173,14 +173,16 @@ static int adf_dsi_vc_send_long(struct dsi_pipe *dsi_pipe, int channel,
 
 	if (intel_dsi->hs) {
 		data_reg = MIPI_HS_GEN_DATA(pipe);
-		mask = HS_DATA_FIFO_FULL;
+		ctrl_reg = MIPI_HS_GEN_CTRL(pipe);
+		mask |= HS_CTRL_FIFO_EMPTY | HS_DATA_FIFO_EMPTY;
 	} else {
 		data_reg = MIPI_LP_GEN_DATA(pipe);
-		mask = LP_DATA_FIFO_FULL;
+		ctrl_reg = MIPI_LP_GEN_CTRL(pipe);
+		mask |= LP_CTRL_FIFO_EMPTY | LP_DATA_FIFO_EMPTY;
 	}
 
-	if (wait_for((REG_READ(MIPI_GEN_FIFO_STAT(pipe)) & mask) == 0, 50))
-		pr_err("Timeout waiting for HS/LP DATA FIFO !full\n");
+	if (wait_for((REG_READ(MIPI_GEN_FIFO_STAT(pipe)) & mask) == mask, 50))
+		pr_err("Timeout waiting for HS/LP DATA FIFO to be empty\n");
 
 	for (i = 0; i < len; i += n) {
 		u32 val = 0;
@@ -193,8 +195,12 @@ static int adf_dsi_vc_send_long(struct dsi_pipe *dsi_pipe, int channel,
 		/* XXX: check for data fifo full, once that is set, write 4
 		 * dwords, then wait for not set, then continue. */
 	}
+	ctrl = len << LONG_PACKET_WORD_COUNT_SHIFT;
+	ctrl |= channel << VIRTUAL_CHANNEL_SHIFT;
+	ctrl |= data_type << DATA_TYPE_SHIFT;
 
-	return adf_dsi_vc_send_short(dsi_pipe, channel, data_type, len);
+	REG_WRITE(ctrl_reg, ctrl);
+	return 0;
 }
 
 static int adf_dsi_vc_write_common(struct dsi_pipe *dsi_pipe,
-- 
1.7.9.5

